Attach kubernetes attributes to traces blogpost (#1454)
open-telemetry/opentelemetry.io: content/en/blog/2022/k8sattributes.md

---

> title: Attaching Kubernetes attributes to traces

Author: Well this is not exclusively about the k8s processor, I mean most of the article is related to that but at the end I mentioned other way to do it using the resource detector processor.

May be just "kubernetes attributes"?

> title: Attaching Kubernetes attributes to traces

Reviewer: @svrnm - thoughts?

> title: Attaching Kubernetes attributes to traces

Reviewer 2: What about "kubernetes metadata"?

---
title: Improved troubleshooting using k8s attributes
linkTitle: Kubernetes attributes
date: 2022-06-23
author: Ruben Vargas
spelling: cSpell:ignore k8sattributes K8sattributes K8sprocessor KUBE
---

Attaching kubernetes resource attributes to the traces is something that is very
useful it help us to excctly identify which resource is failing or having
performance problems. It could be also good to have this information if we want
to correlate traces with other signals like metrics and logs.

In this article you will learn how to configure the OpenTelemetry collector to
use the k8sattributes processor, and we will explore how to configure it in
different scenarios. At the end, we present some alternatives and possible

> future improvements.

Author: All good, except the part of the processor, what I want to mean is some improvements on how to do this in general.  because there is a way to do it with the resource detector and some operator features.

We won’t cover the details of the OpenTelemetry collector pipeline. For that
please, refer to [OpenTelemetry collector documentation](/docs/collector/).

Before diving into the details we need to clarify that the k8s attributes are
not attaching to each single span. Instead the processors attach all the
information to a OpenTelemetry
[Resource](/docs/reference/specification/overview/#resources). this reduces the

> data duplication.

Author: I like this better :) Thank you!

## Using k8sattributes processor

K8sattributes processor is an OpenTelemetry collector processor that attaches
Kubernetes attributes to spans, logs and metrics.

This processor automatically discovers pod attributes and attaches those

> attributes to the spans generated by that pod. If the pod belongs to a

Author: Yeah, well visually it seems to be attached to the span, I mean for example in the Jaeger UI you can see it in the span, but you're right, is on the resource. I'll rephrase this. Thanks!

---
title:
  Improved application & infrastructure troubleshooting attaching k8s attributes

> linkTitle: Attaching Kubernetes attributes to traces

Reviewer: Ok, we're making progress on the title & short title. I'd suggest the following adjustments:

- title: Improved troubleshooting using k8s attributes
- linkTitle: Kubernetes attributes

Thoughts @svrnm @rubenvp8510?

Author: I like, make sense to me

Reviewer 3: sounds good to me!

> date: 2022-06-16

Reviewer: Pls update the date

---
title: Improved troubleshooting using k8s attributes
linkTitle: Kubernetes attributes
date: 2022-06-23
author: Ruben Vargas
spelling: cSpell:ignore k8sattributes K8sattributes K8sprocessor KUBE
---

> Attaching kubernetes resource attributes to traces is something can be very
> useful because it lets you identify which resource (such as a pod) is failing or
> having performance problems. It is also generally useful for correlating across
> other signals.

Reviewer 2: ```suggestion
Attaching kubernetes resource metadata to OpenTelemetr traces is something can be very
useful because it lets you identify which resource (such as a pod) is failing or
having performance problems. It is also generally useful for correlating across
other signals.
```

> In this article, you'll learn how to configure the OpenTelemetry collector to
> use the `k8sattributesprocessor` processor, and we will explore how to configure
> it in different scenarios. At the end, we present some alternatives.

Reviewer 2: ```suggestion
In this article, you'll learn how to configure the OpenTelemetry collector to
use the `k8sattributesprocessor` processor in different scenarios.
At the end, we present some alternatives.
```

          thrift_http:
      otlp:
        protocols:
          grpc:
          http:

    processors:
      resourcedetection:
        detectors: [env]
        timeout: 2s
        override: false

    exporters:
      jaeger:
        endpoint: jaeger-all-in-one-collector:14250
        tls:
          insecure: true

    service:
      pipelines:
        traces:
          receivers: [otlp, jaeger]
          processors: [resourcedetection]
          exporters: [jaeger]
```

And then deploy the vertx app example, you can see the `OTEL_RESOURCE_ATTRIBUTES`
environment variable was injected with some values in the sidecar container. Some of them use the Kubernetes downward API to get the attribute values.

An example of the value of the env var:

```yaml
- name: OTEL_RESOURCE_ATTRIBUTES
  value: k8s.deployment.name=dep-vertex,k8s.deployment.uid=ef3fe26b-a690-4746-9119-d2dbd94b469f,
  k8s.namespace.name=default,k8s.node.name=$(OTEL_RESOURCE_ATTRIBUTES_NODE_NAME),k8s.pod.name=
  (OTEL_RESOURCE_ATTRIBUTES_POD_NAME),k8s.pod.uid=$(OTEL_RESOURCE_ATTRIBUTES_POD_UID),k8s.replicaset
  name=dep-vertex-59b6f76585,k8s.replicaset.uid=5127bc38-e298-40e1-95df-f4a777e3176c
```

## Conclusion

> We have learned about how to do a basic configuration that allows the
> OpenTelemetry collector to attach Kubernetes resource attributes to traces.
> There could be other complex scenarios but I hope this could serve as a base
> that allows the reader to understand how the to add Kubernetes attributes to the
> spans generated by the pods. We explored how to configure the k8sattributes
> operator in order to accomplish this, and saw some improvements introduced on
> the OpenTelemetry operator that allow us to do this in a simple way on sidecar
> mode

Reviewer 2: I struggled to do minor copy-edits on this one, so this is kind of a rewrite of the paragraph, but with some stuff stripped out.

```suggestion
This post covers how to configure the OpenTelemetry collector to attach
Kubernetes resource metadata as resource attributes to OpenTelemetry traces.
The scenarios covered, although basic, illustrate how to add this kind of metadata
to traces so that you can incorporate the technique into other more sophisticated
scenarios.
```

I have half a mind to just suggest dropping the conclusion. I tend not to like conclusions that just summarize what was already covered without an explicit call to action or place to go learn more about a topic. What do you think?

Author: I agree, may be I should add a paragraph about how you can learn more about other processor options and different  configurations?

The processor internally maintains a list of pods and an associated attribute,
usually the IP address of the pod, and uses this attribute to know which pod
generates a certain span.

![k8sattributes processor data flow](/img/blog-k8sattributes/k8sprocessor.png)

In the figure above you can see how the data flows: The table of pods is fetched
using Kubernetes API, while the pod IP is extracted from the connection context
between the pod and the collector.

The `k8sattributesprocessor` can work in different modes, and it depends on how
the collector is configured. Let's explore one common scenario, when the
collector is deployed as daemonset.

### DaemonSet mode

Let’s take a look at how we can configure the collector in daemonset mode, known
as an agent mode in the k8sattributes documentation.

When we deploy the collector as the DaemonSet we have one collector pod per
node. We need to configure the collector service account to have permissions to
fetch all pods information. In order to do that, we will create a `ClusterRole`
with the necessary permissions.

> Here are the minimum permissions required for make k8sattributes processor
> works:

Reviewer 2: ```suggestion
Here are the minimum permissions required to make the `k8sattributesprocessor`
work:
```

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: otel-collector
rules:
  - apiGroups: [""]
    resources: ["pods", "namespaces"]
    verbs: ["get", "watch", "list"]
```

> Now, we are going to deploy the collector in daemonset mode. It is recommended
> to set a filter to only fetch the pods that belong to the node in which the
> collector is deployed. This is because if we are on a large cluster, we don’t
> want to maintain a huge list of pods.

Reviewer 2: ```suggestion
Next, deploy the collector in daemonset mode. It is recommended
to set a filter to only fetch the pods that belong to the node in which the
collector is deployed. This is because if you have a large cluster, you don’t
want to maintain a huge list of pods.
```

This is the manifest used in this blog to show how the processor works:

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: otel-collector-daemonset
spec:
  mode: daemonset
  image: ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector-contrib:0.47.0
  serviceAccount: attributes-account
  env:
    - name: KUBE_NODE_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
  config: |
    receivers:
      jaeger:
          protocols:
              grpc:
              thrift_binary:
              thrift_compact:
              thrift_http:
      otlp:
          protocols:
              grpc:
              http:

    processors:
         k8sattributes:
             filter:
                 node_from_env_var: KUBE_NODE_NAME
    exporters:
      jaeger:
        endpoint: jaeger-all-in-one-collector:14250
        tls:
          insecure: true

    service:
      pipelines:
        traces:
          receivers: [otlp, jaeger]
          processors: [k8sattributes]
          exporters: [jaeger]
```

> The main parts to note are that we are using the contrib collector image, the
> reason is because this processor is not part of the OpenTelemetry collector
> core. Other things to notice are the filter mentioned above, and the use of a
> specific service account which is created previously and contains the
> permissions to fetch the pod list.

Reviewer 2: ```suggestion
The main parts to note are that it uses the contrib collector image. The
`k8sattributesprocessor` is not part of the OpenTelemetry collector
core, but the contrib distribution has it. Other things to notice are the filter
mentioned above, and the use of a previously-created specific service account,
which contains the permissions to fetch the pod list.
```

> Now, we can deploy the manifest above and the vertex app example to generate
> some traces.

Reviewer 2: ```suggestion
Next, deploy the manifest and the vertex app example to generate
some traces.
```

![Jaeger UI showing the span attributes](/img/blog-k8sattributes/jaeger-k8sattributes.png)

As we can see, each span of the trace now has the corresponding pod attributes
attached to it.

> We can restrict the above configuration to a certain namespace if we add the
> namespace on the k8sattributes processor filter like this:

Reviewer 2: ```suggestion
You can restrict the above configuration to a certain namespace if you add the
namespace on the `k8sattributesprocessor` filter like this:
```

```yaml
processors:
  k8sattributes:
    filter:
      namespace: my_namespace
```

> In this way we can create a `Role` and don’t need to create a `ClusterRole`,
> reducing the scope of the collector service account to a single namespace.

Reviewer 2: ```suggestion
In this way you can create a `Role` and don’t need to create a `ClusterRole`,
reducing the scope of the collector service account to a single namespace.
```

## Using Resource detector processor

Recently the OpenTelemetry operator incorporated a new feature introduced in
version v0.50.0, see
[PR #832](https://github.com/open-telemetry/opentelemetry-operator/pull/832).

> This feature sets the `OTEL_RESOURCE_ATTRIBUTES` environment variable on the
> collector container with the k8s pod attributes. It allows us to use the
> resource detector processor, which attaches the environment variable values to
> the spans. This only works when the collector is deployed in sidecar mode.

Reviewer 2: ```suggestion
This feature sets the `OTEL_RESOURCE_ATTRIBUTES` environment variable on the
collector container with the k8s pod attributes. It lets you to use the
resource detector processor, which attaches the environment variable values to
the spans. This only works when the collector is deployed in sidecar mode.
```

> For example if we deploy this manifest:

Reviewer 2: ```suggestion
For example if you deploy this manifest:
```

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: sidecar-for-my-app
spec:
  mode: sidecar
  image: ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector-contrib:0.47.0
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
          thrift_binary:
          thrift_compact:
          thrift_http:
      otlp:
        protocols:
          grpc:
          http:

    processors:
      resourcedetection:
        detectors: [env]
        timeout: 2s
        override: false

    exporters:
      jaeger:
        endpoint: jaeger-all-in-one-collector:14250
        tls:
          insecure: true

    service:
      pipelines:
        traces:
          receivers: [otlp, jaeger]
          processors: [resourcedetection]
          exporters: [jaeger]
```

> And then we deploy the vertx app example, we can see in the sidecar container,
> the `OTEL_RESOURCE_ATTRIBUTES` was injected with some values, some of them are
> using the Kubernetes downward API to get the attribute values.

Reviewer 2: ```suggestion
And then deploy the vertx app example, you can see the `OTEL_RESOURCE_ATTRIBUTES`
environment variable was injected with some values in the sidecar container. Some of them use the Kubernetes downward API to get the attribute values.
```

An example of the value of the env var:

```yaml
- name: OTEL_RESOURCE_ATTRIBUTES
  value: k8s.deployment.name=dep-vertex,k8s.deployment.uid=ef3fe26b-a690-4746-9119-d2dbd94b469f,
  k8s.namespace.name=default,k8s.node.name=$(OTEL_RESOURCE_ATTRIBUTES_NODE_NAME),k8s.pod.name=
  (OTEL_RESOURCE_ATTRIBUTES_POD_NAME),k8s.pod.uid=$(OTEL_RESOURCE_ATTRIBUTES_POD_UID),k8s.replicaset
  name=dep-vertex-59b6f76585,k8s.replicaset.uid=5127bc38-e298-40e1-95df-f4a777e3176c
```

> ## Conclusions

Reviewer 2: ```suggestion
## Conclusion
```

---
title: Improved troubleshooting using k8s attributes
linkTitle: Kubernetes attributes
date: 2022-06-27
author: Ruben Vargas
spelling:
  cSpell:ignore k8sattributes k8sattributesprocessor K8sattributes k8sprocessor
  cSpell:ignore K8sprocessor KUBE
---

> Attaching Kubernetes resource metadata to OpenTelemetry traces can be very

Reviewer 3: replace
> can be very useful because it lets you identify
with 
> is useful to identify

useful because it lets you identify which resource (such as a pod) is failing or

> having performance problems. It is also generally useful for correlating across

Reviewer 3: > It is also generally useful for correlating across other signals.

Remove the "generally"

Can you provide an example?

other signals.

In this article, you'll learn how to configure the OpenTelemetry collector to

> use the `k8sattributesprocessor` processor in different scenarios. At the end,

Reviewer 3: replace `k8sattributesprocessor` with a link to it (e.g. https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor)

> we present some alternatives.

Reviewer: > At the end, we present some alternatives.

Really?

Author: May be is just a short mention, but we present an alternative of using the resource detector processor.

Reviewer: I'd propose dropping that sentence.

Details of the OpenTelemetry collector pipeline won't be covered in this post.
For those details please, refer to
[OpenTelemetry collector documentation](/docs/collector/).

## How k8s attributes are attached

At a high level, k8s attributes are attached to traces as
[resources](/docs/concepts/glossary/#resource). This is for two reasons:

> 1.  K8s attributes fit the definition of what a resource is -- an entity for

Reviewer 3: replace "--" with ": "

    which telemetry is recorded
2.  It centralizes this metadata, which is relevant for any generated span.

Let's dive in and see how to do it!

## Using k8sattributes processor

> The `k8sattributesprocessor` is an OpenTelemetry collector processor that

Reviewer 3: this sentence is redundant, since the next sentence says the same thing, you might remove it or merge them

> The `k8sattributesprocessor` is an OpenTelemetry collector processor that automatically discovers pod attributes and attaches those ...

attaches Kubernetes attributes to spans, logs and metrics.

This processor automatically discovers pod attributes and attaches those
attributes to a resource associated with the spans generated by that pod. If the
pod belongs to a `Deployment` or a `ReplicaSet`, it will also discover it's
attributes.

Some attributes we can attach to the resource are:

- Node name `k8s.node.name`
- Pod name `k8s.pod.name`
- Pod UID `k8s.pod.uid`
- Namespace `k8s.namespace.name`
- Deployment name, if the pod was created by a deployment

Such attributes adhere to OpenTelemetry semantic conventions. For details, see
the [Kubernetes resource semantic conventions][].

The processor internally maintains a list of pods and an associated attribute,
usually the IP address of the pod, and uses this attribute to know which pod
generates a certain span.

> ![k8sattributes processor data flow](/img/blog-k8sattributes/k8sprocessor.png)

Reviewer: In the diagram, "Add attributes" names an action. IMHO, it doesn't make sense to have an entity in this diagram representing an action. Or am I misunderstanding something here?

Author: I can't find a better way to represent it :/

Reviewer: Consider just dropping the "Add attributes" box. The figure is meant to represent "how the data flows".

  serviceAccount: attributes-account
  env:
    - name: KUBE_NODE_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
  config: |
    receivers:
      jaeger:
          protocols:
              grpc:
              thrift_binary:
              thrift_compact:
              thrift_http:
      otlp:
          protocols:
              grpc:
              http:

    processors:
         k8sattributes:
             filter:
                 node_from_env_var: KUBE_NODE_NAME
    exporters:
      jaeger:
        endpoint: jaeger-all-in-one-collector:14250
        tls:
          insecure: true

    service:
      pipelines:
        traces:
          receivers: [otlp, jaeger]
          processors: [k8sattributes]
          exporters: [jaeger]
```

The main parts to note are that it uses the contrib collector image. The
`k8sattributesprocessor` is not part of the OpenTelemetry collector core, but
the contrib distribution has it. Other things to notice are the filter mentioned
above, and the use of a previously-created specific service account, which
contains the permissions to fetch the pod list.

> Next, deploy the manifest and the vertex app example to generate some traces.

Reviewer: Can you provide a link to the "vertex app example"?

          thrift_http:
      otlp:
        protocols:
          grpc:
          http:

    processors:
      resourcedetection:
        detectors: [env]
        timeout: 2s
        override: false

    exporters:
      jaeger:
        endpoint: jaeger-all-in-one-collector:14250
        tls:
          insecure: true

    service:
      pipelines:
        traces:
          receivers: [otlp, jaeger]
          processors: [resourcedetection]
          exporters: [jaeger]
```

And then deploy the vertex app example, you can see the
`OTEL_RESOURCE_ATTRIBUTES` environment variable gets injected with some values
in the sidecar container. Some of them use the Kubernetes downward API to get
the attribute values.

Here's an example of the value of the environment variable:

```yaml
- name: OTEL_RESOURCE_ATTRIBUTES
  value: k8s.deployment.name=dep-vertex,k8s.deployment.uid=ef3fe26b-a690-4746-9119-d2dbd94b469f,
  k8s.namespace.name=default,k8s.node.name=$(OTEL_RESOURCE_ATTRIBUTES_NODE_NAME),k8s.pod.name=
  (OTEL_RESOURCE_ATTRIBUTES_POD_NAME),k8s.pod.uid=$(OTEL_RESOURCE_ATTRIBUTES_POD_UID),k8s.replicaset
  name=dep-vertex-59b6f76585,k8s.replicaset.uid=5127bc38-e298-40e1-95df-f4a777e3176c
```

> ## Conclusion
> 
> This post covers how to configure the OpenTelemetry collector to attach
> Kubernetes resource metadata as resource attributes to OpenTelemetry traces. The
> scenarios covered, although basic, illustrate how to add this kind of metadata
> to traces so that you can incorporate the technique into other more
> sophisticated scenarios. If you want to learn more about different scenarios or
> options for configure the processors you can see the K8sattributes processor
> documentation where you can find more scenarios like sidecar, or when one
> collector as an agent report to another collector.
> 
> ## References
> 
> - [K8sattributes processor](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor)
> - [K8sattributes processor RBAC](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor#hdr-RBAC)
> - [OpenTelemetry Kubernetes attributes](/docs/reference/specification/resource/semantic_conventions/k8s)
> - [resource detector processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/README.md)
> 
> [pr#832]: https://github.com/open-telemetry/opentelemetry-operator/pull/832
> [opentelemetry operator]:
>   https://github.com/open-telemetry/opentelemetry-operator
> [kubernetes resource semantic conventions]:
>   /docs/reference/specification/resource/semantic_conventions/k8s

Reviewer: @rubenvp8510 and @cartermp : what do you think of the suggestion below as an alternative to a "conclusion"

```suggestion
## Learn more

This post covers how to configure the OpenTelemetry collector to attach
Kubernetes resource metadata as resource attributes to OpenTelemetry traces. The
scenarios covered, although basic, illustrate how to add this kind of metadata
to traces so that you can incorporate the technique into other more
sophisticated scenarios. If you want to learn more about different scenarios or
options for configure the processors you can see the [K8sattributes processor
documentation][] where you can find more scenarios like sidecar, or when one
collector as an agent report to another collector.

## References

- [K8sattributes processor](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor)
- [K8sattributes processor RBAC](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor#hdr-RBAC)
- [OpenTelemetry Kubernetes attributes](/docs/reference/specification/resource/semantic_conventions/k8s)
- [Resource detector processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/README.md)

[pr#832]: https://github.com/open-telemetry/opentelemetry-operator/pull/832
[opentelemetry operator]:
  https://github.com/open-telemetry/opentelemetry-operator
[k8sattributes processor documentation]:
  https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor
[kubernetes resource semantic conventions]:
  /docs/reference/specification/resource/semantic_conventions/k8s
```

Author: I like it.
